valhalla/midgard/point_tile_index.h

```diff
@@ -0,0 +1,135 @@
+#pragma once
+
+#include <list>
+#include <unordered_set>
+
+#include "midgard/linesegment2.h"
+#include "midgard/pointll.h"
+
+namespace valhalla {
+
+namespace midgard {
+
+/**
+ * Indexes a given container_t of PointLL's into a tiled space.
+ * (So yea, I'm using the word "tile" which is a heavily overloaded term.
+ * Maybe "pixel" is a better term? Or "grid"? Or "discretized space"?)
+ *
+ * Basically, I've taken the lat/lon ranges for the earth and subdivided them
+ * into smaller and smaller rectangles until we reach a rectangle size that
+ * only just big enough to meet the given tile_width_degrees.
+ *
+ * Then, in a single pass in the ::tile() method, we compute each PointLL's
+ * "TileId" - which is just a Cartesian coordinate into the tiled space.
+ * We then place each PointLL into its appropriate TileId bin.
+ *
+ * Once all the points are binned we can quickly solve problems like "what
+ * points are near me" - in effectively O(1) time (the speed of a hash
+ * lookup).
+ *
+ * Of important note is the fact that to determine the "points within an
+ * exact distance of a given point" you must round up all points in all
+ * surrounding tiles. You can subsequently compute the distance bewteen your
+ * point and each "nearby" point using a more accurate distance computation
+ * of your choosing.
+ */
+class PointTileIndex {
+  // A coordinate into our tile/grid space. Using uint32_t means we could
+  // theoretically subdivide 360 degrees of latitude into 2^32 bins. Very
+  // roughly, that's about 1e-8 degrees or 1 cm.
+  struct TileId {
+    uint32_t x, y;
+    friend bool operator==(const TileId& L, const TileId& R) {
+      return L.x == R.x && L.y == R.y;
+    }
+  };
+
+  // prev/next "iterators" to ensure we stay on the tiled-space when
+  // we might otherwise underflow/overflow.
+  uint32_t prevx(uint32_t x) {
+    return (x > 0) ? x - 1 : num_x_subdivisions - 1;
+  }
+  uint32_t nextx(uint32_t x) {
+    return (x < num_x_subdivisions - 1) ? x + 1 : 0;
+  }
+  uint32_t prevy(uint32_t y) {
+    return (y > 0) ? y - 1 : num_y_subdivisions - 1;
+  }
+  uint32_t nexty(uint32_t y) {
+    return (y < num_y_subdivisions - 1) ? y + 1 : 0;
+  }
+
+  struct TileId_hash_functor {
+    size_t operator()(const TileId& tid) const {
+#if 0
```

Comment:
What is the intent of this if block?

Reply:
Apparently someone in valhalla's past has thought about ways to hash-combine (x,y)'s as a means for well-distributed (performant) hashing.  I simply wanted to point out that I did try using it but found that it wasn't as performant as my approach.

Happy to remove it... I'll simply add a comment to the same effect.